src/ClpIrV1Decoder.cpp

```diff
@@ -0,0 +1,169 @@
+#include "ClpIrV1Decoder.hpp"
+
+#include <emscripten/emscripten.h>
+#include <emscripten/val.h>
+
+#include <clp/ffi/encoding_methods.hpp>
+#include <clp/ffi/ir_stream/decoding_methods.hpp>
+#include <clp/TimestampPattern.hpp>
+
+#include "spdlog/spdlog.h"
+#include "types.hpp"
+
+static constexpr size_t cDefaultInitialNumLogEvents = 500'000;
+static constexpr size_t cFullRangeEndIdx = 0;
+static constexpr size_t cLogLevelNone = 0;
+
+auto ClpIrV1Decoder::create(emscripten::val const& data_array) -> ClpIrV1Decoder* {
+    auto length = data_array["length"].as<size_t>();
+    SPDLOG_INFO("ClpIrV1Decoder::ClpIrV1Decoder() got buffer of length={}", length);
+
+    auto data_buffer = std::vector<char const>(length);
+    emscripten::val::module_property("HEAPU8")
+            .call<void>("set", data_array, reinterpret_cast<uintptr_t>(data_buffer.data()));
+
+    auto zstd_decompressor = std::make_shared<clp::streaming_compression::zstd::Decompressor>();
+    zstd_decompressor->open(data_buffer.data(), length);
+
+    bool is_four_bytes_encoding{true};
+    if (auto const err{
+                clp::ffi::ir_stream::get_encoding_type(*zstd_decompressor, is_four_bytes_encoding)
+        };
+        clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success != err)
+    {
+        SPDLOG_CRITICAL("Failed to decode encoding type.");
+        throw err;
+    }
+    if (false == is_four_bytes_encoding) {
+        SPDLOG_CRITICAL("Is not four byte encoding.");
+        throw std::exception();
```

Comment:
@LinZhihao-723

Shall we use a custom class of exception here? Note that with "Release" (`-O3`) enabled, exception can be caught on the JS side but the exact messages / information would not be available. With that said, in the future we can enable exception once we integrate with the Log Viewer (if we determine the overhead for exception handling is negligible).

Reply:
Discussed offline. If we need to throw an expectation, we should derive an exception class from `clp::TraceableException`